caikit/runtime/client/remote_model_finder.py

```diff
@@ -195,110 +238,139 @@ def _safe_discover(self) -> Dict[str, str]:
 
-        # Load the response as a json object
-        model_info = resp.json()
+            # Send HTTP Request
+            try:
+                resp = session.get(target)
+
+                if resp.status_code != 200:
+                    log.warning(
+                        "Unable to discover modules from remote: %s. Error: %s",
+                        target,
+                        resp.reason,
+                    )
+                else:
+
+                    # Load the response as a json object
+                    model_info = resp.json()
+
+                    # Parse response into dictionary of name->id
+                    for model_dict in model_info.get("models", []):
+                        model_name = model_dict.get("name")
+                        module_id = model_dict.get("module_id")
+
+                        log.debug(
+                            "Discovered model %s with module_id %s from remote runtime",
+                            model_name,
+                            module_id,
+                        )
+                        # NOTE: If multiple servers support the same model, the last to
+                        #   be checked will win
+                        supported_modules[model_name] = ModuleConnectionInfo(
+                            conn, module_id
+                        )
+            except RequestException as exc:
+                log.warning(
+                    "Unable to discover modules from remote: %s. Error: %s",
+                    target,
+                    str(exc),
+                )
 
-        # Parse response into dictionary of name->id
-        supported_modules = {}
-        for model_dict in model_info.get("models", []):
-            model_name = model_dict.get("name")
-            module_id = model_dict.get("module_id")
+        return supported_modules
 
-            log.debug(
-                "Discovered model %s with module_id %s from remote runtime",
-                model_name,
-                module_id,
+    def _render_conn_template(self, model_name: str) -> Optional[ConnectionInfo]:
+        """Common utility to get the connection for a given model"""
+        if self._connection_template is not None:
+            conn_dict = self._connection_template.to_dict()
+            conn_dict["hostname"] = self._connection_template.hostname.format(
+                model_name
             )
-            supported_modules[model_name] = module_id
-
-        return supported_modules
+            return ConnectionInfo(**conn_dict)
+
+    def _get_conn_candidates(self, model_name: Optional[str]) -> List[ConnectionInfo]:
+        """Common utility to get all connections to try"""
+        candidate_conns = list(self._connections.values())
+        if (
+            model_name is not None
+            and self._connection_template is not None
+            and (model_conn := self._render_conn_template(model_name))
+        ):
+            candidate_conns.append(model_conn)
```

Comment:
this will put the connection with template at the back of the list so when trying to make connection request to check for model, for example [here](https://github.com/caikit/caikit/pull/681/files#diff-b2d06d99d6692ffc8747bc7bc99c53f4ae9122c17e27971babf99052c82f9141R307), we would end up making a call to incorrect server first (assuming it satisfies the template) right?  Should we prepend instead?

Reply:
Hmm, good call. Probably best to prepend. This does touch on one open question I have: how to handle multiple servers supporting the same model? Currently, it's [last one wins](https://github.com/caikit/caikit/pull/681/files#diff-b2d06d99d6692ffc8747bc7bc99c53f4ae9122c17e27971babf99052c82f9141R341), so moving to prepending would not actually save any effort since all servers would be checked anyway. We could certainly make this more efficient, though! That said, this is a load-time operation, so once the connection is established, it should have no effect on the inference latency.

Reply:
true. But:
1. Wouldn't it be unintuitive for the last one to win, because one would expect first one to get discovered would win and have first come first serve 
2. Making call to incorrect backend would also create noise, so from the perspective of debugging a routing / model loading issue, we would see more failures from servers that are not responsible.

Reply:
Updated so that the first-in wins. The catch is the the connection calls are not per-model since models are `1:N` with runtimes, so we can't really limit the discovery calls by model.

Reply:
true. 

Thanks for updating